traffic_ops/traffic_ops_golang/auth/authenticate.go

```diff
@@ -0,0 +1,211 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"crypto/rand"
+	"crypto/subtle"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"golang.org/x/crypto/scrypt"
+)
+
+// SCRYPTComponents the input parameters to the Scrypt encryption key format
+type SCRYPTComponents struct {
+	Algorithm string // The SCRYPT algorithm prefix
+	N         int    // CPU/memory cost parameter (logN)
+	R         int    // block size parameter (octets)
+	P         int    // parallelization parameter (positive int)
+	Salt      []byte // salt value
+	SaltLen   int    // bytes to use as salt (octets)
+	DK        []byte // derived key value
+	DKLen     int    // length of the derived key (octets)
+}
+
+const KEY_DELIM = ":"
+
+// The SCRYPT functionality defined in this package is derived based upon the following
+// references:
+// https://godoc.org/golang.org/x/crypto/scrypt
+// https://www.tarsnap.com/scrypt/scrypt.pdf
+var DefaultParams = SCRYPTComponents{
+	Algorithm: "SCRYPT",
+	N:         16384,
+	R:         8,
+	P:         1,
+	SaltLen:   16,
+	DKLen:     64}
+
+// DerivePassword uses the https://godoc.org/golang.org/x/crypto/scrypt package to
+// return an encrypted password that is compatible with the
+// Perl CPAN library Crypt::ScryptKDF for backward compatibility
+// to authenticate through the Perl API the same way.
+// See: http://cpansearch.perl.org/src/MIK/Crypt-ScryptKDF-0.010/lib/Crypt/ScryptKDF.pm
+func DerivePassword(password string) (string, error) {
+	var salt []byte
+	var err error
+	salt, err = generateSalt(DefaultParams.DKLen)
+	if err != nil {
+		return "", errors.New("generating salt: " + err.Error())
+	}
+	key, err := scrypt.Key([]byte(password), salt, DefaultParams.N, DefaultParams.R, DefaultParams.P, DefaultParams.DKLen)
+	if err != nil {
+		return "", err
+	}
+	nStr := strconv.Itoa(DefaultParams.N)
+	if err != nil {
+		return "", errors.New("converting N: " + err.Error())
+	}
+	rStr := strconv.Itoa(DefaultParams.R)
+	pStr := strconv.Itoa(DefaultParams.P)
+	saltBase64 := base64.StdEncoding.EncodeToString(salt)
+	keyBase64 := base64.StdEncoding.EncodeToString(key)
+
+	// The SCRYPT prefix is added because the Mojolicious Perl library adds this as a
+	// prefix to every password in the database.  So it's added for compatibility.
+	scryptPass := []string{DefaultParams.Algorithm, nStr, rStr, pStr, saltBase64, keyBase64}
+
+	return strings.Join(scryptPass, KEY_DELIM), nil
+}
+
+// VerifyPassword parses the original Derived Key (DK) from the SCRYPT password
+// so that it can compare that with the password/scriptPassword param
+func VerifyPassword(password string, scryptPassword string) (bool, error) {
```

Comment:
I had to read the whole function to figure out what cases return true/false, and what return error/nil, and when `(false,nil)` and `(true,error)` were returned. As-is, adding the specifics about what's returned in what case to the function-level comment would be really helpful. 

But, I like the original better, just returning an error, where `nil` means success. Since `(true,error)` never makes sense, we can return `nil` on success. Just change the last line to `return errors.New("invalid password")`.

But, it works, so I'll pass the review, if you disagree.

Reply:
@rob05c I went back and forth on the (bool, error) return.  I decided to go back to the original and added the return error that you mentioned.